src/server/saved_objects/serialization/index.ts

```diff
@@ -57,88 +58,108 @@ export interface SavedObjectDoc {
+  updated_at?: Date;
 
   [rootProp: string]: any;
 }
 
-/**
- * Determines whether or not the raw document can be converted to a saved object.
- *
- * @param {RawDoc} rawDoc - The raw ES document to be tested
- */
-export function isRawSavedObject(rawDoc: RawDoc) {
-  const { type } = rawDoc._source;
-  return type && rawDoc._id.startsWith(type) && rawDoc._source.hasOwnProperty(type);
-}
-
-/**
- * Converts a document from the format that is stored in elasticsearch to the saved object client format.
- *
- *  @param {RawDoc} rawDoc - The raw ES document to be converted to saved object format.
- */
-export function rawToSavedObject({ _id, _source, _version }: RawDoc): SavedObjectDoc {
-  const { type } = _source;
-  return {
-    type,
-    id: trimIdPrefix(type, _id),
-    attributes: _source[type],
-    ...(_source.migrationVersion && { migrationVersion: _source.migrationVersion }),
-    ...(_source.updated_at && { updated_at: _source.updated_at }),
-    ...(_version != null && { version: _version }),
-  };
-}
-
-/**
- * Converts a document from the saved object client format to the format that is stored in elasticsearch.
- *
- * @param {SavedObjectDoc} savedObj - The saved object to be converted to raw ES format.
- */
-export function savedObjectToRaw(savedObj: SavedObjectDoc): RawDoc {
-  const { id, type, attributes, version } = savedObj;
-  const source = {
-    ...savedObj,
-    [type]: attributes,
-  };
-
-  delete source.id;
-  delete source.attributes;
-  delete source.version;
-
-  return {
-    _id: generateRawId(type, id),
-    _source: source,
-    _type: ROOT_TYPE,
-    ...(version != null && { _version: version }),
-  };
-}
-
-/**
- * Given a saved object type and id, generates the compound id that is stored in the raw document.
- *
- * @param {string} type - The saved object type
- * @param {string} id - The id of the saved object
- */
-export function generateRawId(type: string, id?: string) {
-  return `${type}:${id || uuid.v1()}`;
-}
-
 function assertNonEmptyString(value: string, name: string) {
   if (!value || typeof value !== 'string') {
     throw new TypeError(`Expected "${value}" to be a ${name}`);
   }
 }
 
-function trimIdPrefix(type: string, id: string) {
-  assertNonEmptyString(id, 'document id');
-  assertNonEmptyString(type, 'saved object type');
+export class SavedObjectsSerializer {
+  private readonly schema: SavedObjectsSchema;
 
-  const prefix = `${type}:`;
+  constructor(schema: SavedObjectsSchema) {
+    this.schema = schema;
+  }
+  /**
+   * Determines whether or not the raw document can be converted to a saved object.
+   *
+   * @param {RawDoc} rawDoc - The raw ES document to be tested
+   */
+  public isRawSavedObject(rawDoc: RawDoc) {
+    const { type, namespace } = rawDoc._source;
+    const namespacePrefix =
+      namespace && !this.schema.isNamespaceAgnostic(type) ? `${namespace}:` : '';
+    return (
+      type &&
+      rawDoc._id.startsWith(`${namespacePrefix}${type}`) &&
```

Comment:
I think this needs a trailing colon to be precise:

```
rawDoc._id.startsWith(`${namespacePrefix}${type}:`)
```

... This was a bug in my original implementation.